src/xdg.c

```diff
@@ -340,10 +340,17 @@ xdg_toplevel_view_map(struct view *view)
 		 * view_set_fullscreen/view_maximize() below). "Current"
 		 * dimensions remain zero until handle_commit().
 		 */
-		view->pending.width = xdg_surface->current.geometry.width;
-		view->pending.height = xdg_surface->current.geometry.height;
+		if (wlr_box_empty(&view->pending)) {
+			view->pending.width =
+				xdg_surface->current.geometry.width;
+			view->pending.height =
+				xdg_surface->current.geometry.height;
+		}
+
+		if (view_is_floating(view)) {
+			position_xdg_toplevel_view(view);
```

Comment:
I think the `view_center()` call in `position_xdg_toplevel_view()` will overwrite `view->pending`

Reply:
That's intentional in the case of floating views, because we really do want to change the geometry in that case.

Reply:
The earlier `wlr_box_empty()` is just a safeguard to ensure we don't undo some previous, possibly still pending, geometry change and overwrite it with the current surface geometry (which may be out of date).

Reply:
Right, but as this is the first map call there should be no pending changes?

Reply:
And position_xdg_toplevel only sets x and y of pending, based on the width and height of the surface.
Which should then be used as natural geometry once we call view_maximize() (within the first map branch).

Reply:
view_set_decoration() however, will call `apply_xxx_geometry()` and in this case the `apply_maximized_geometry()` variant (due to the earlier call to request_maximize before the map). I am missing a link on how any of this relates to `view->pending`.

Reply:
`handle_request_maximize()` or `handle_request_fullscreen()` could be called before first map and in that case would have updated `view->pending` already.

Reply:
Alright, after some much needed sleep and some more debugging this all starts to make sense now:
- `wlr_box_empty()` does nothing to solve the actual issue here because the sizes are the same in this case, e.g. whatever the pre-map maxmize_request sent to the client via configure.
- `view_center()` calculates the wrong result, it indeed uses the ssd border thickness but then does not add it afterwards (which is required because `view_move()` addresses the actual surface, not the SSD (which is using negative positions in relation to the view tree)
- `position_xdg_toplevel_view()` "fixes" that by manually modifying `view->current.{x,y}` but does <b>not</b> call `view_move()` which causes the coordinates to be out of sync.
- Using `view_move()` in `position_xdg_toplevel_view()` instead also fixes the issue but is a bad solution as we don't have to call it in the first place (so this PR does the absolutely correct thing in that regard).
- This btw also results in usual views not actually being quite centered when spawned, but that is something for another PR

Reply:
Nice analysis. Writing directly to view->current did look "sketchy" to me but I didn't look at it further. I think xwayland.c has some similar code which might also be wrong in that case.